docs: Terraform style guide (#34876)
hashicorp/terraform: website/docs/language/style.mdx

## Code Style

Writing Terraform code in a consistent style makes it easier to read and maintain. This section reviews code style patterns that we recommend that you adopt.

- Run `terraform fmt` and `terraform validate` before committing your code to source control.
- Use `#` for single and multi-line comments.
- Use nouns for resource names and do not include the resource type in the name. 
- Use underscores to separate multiple words in a resource name. Wrap resource type and name in double quotes in your resource definition.
- Define dependent resources after those they depend on so that your code builds on itself.
- Include a type and description for every variable.
- Include a default value for optional variables.
- Include a value and description for every output.
- Use descriptive nouns to name your input variables, outputs, and local values and use underscores to separate multiple words. 
- Avoid overuse of variables and local values.
- Always include a default provider configuration.

> - Use `count` and `for_each` sparingly.

Reviewer: I'm just a user, but this line is really ambiguous and vague. "Sparingly" implies rarely, if ever, use them, but I'm sure the majority of the community would agree on the exact opposite, and that the best practice is not to hard-code 10 of the exact same resources individually, which is what the style guide would imply if you say to use for_each "sparingly".

Does HCP really want to imply we should hard-code each individual resource instead of using for_each with a var?

We recommend that you run `terraform fmt` before each commit to source control. You can use mechanisms such as [Git pre-commit hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks) to automatically run this command each time you commit your code.

If you use Microsoft VS Code, we recommend that you use the [Terraform VS Code extension](https://marketplace.visualstudio.com/items?itemName=HashiCorp.terraform). The Terraform extension enables syntax highlighting, syntax validation, automatic code formatting and many more features like integration with Terraform Cloud.
If your development environment or text editor supports the [Language Server Protocol](https://microsoft.github.io/language-server-protocol/), we recommend that you use the [Terraform Language Server](https://github.com/hashicorp/terraform-ls). This provides a near identical featureset of our VS Code extension.

## Code validation

> The `terraform validate` command checks that your code is syntactically valid and internally consistent, regardless of any provided variables or existing state. The `validate` command does not check if argument values are valid for a specific provider, but it will verify that they are the correct type.

Reviewer 2: ```suggestion
The `terraform validate` command checks that your configuration is syntactically valid and internally consistent. The `validate` command does not check if argument values are valid for a specific provider, but it will verify that they are the correct type. It does not evaluate any existing state.
```

> The `terraform validate` command is safe to run automatically and frequently. You can configure your text editor to run this command as a post-save check, define it as a pre-commit hook in the Git repository, or run it as a step in a CI/CD pipeline.

Reviewer 2: ```suggestion
The `terraform validate` command is safe to run automatically and frequently. You can configure your text editor to run this command as a post-save check, define it as a pre-commit hook in a Git repository, or run it as a step in a CI/CD pipeline.
```

For more information, refer to the [Terraform `validate` documentation](/terraform/cli/commands/validate).

## File names

We recommend the following file naming conventions:

> - `backend.tf`: Contains your [backend configuration](/terraform/language/settings/backends/configuration). You can define multiple `terraform` blocks in your configuration to separate your backend configuration from your Terraform and provider versioning configuration. 

Reviewer 2: backticks in a list render as links. is there a word we can add or edit the sentence?

(i think we had a ticket with webdev to fix this someday)

## Provider aliasing

Provider aliasing lets you define multiple `provider` blocks for the same Terraform provider. For example, this is useful if you are using a provider that can only target a single region, but you want to deploy to multiple regions. The `provider` meta-argument for resources and the `providers` meta-argument for modules lets you specify which provider to use.

<CodeBlockConfig filename="providers.tf" hideClipboard>

```hcl
# The default provider configuration; resources that begin with `aws_` will use
# it as the default, and it can be referenced as `aws`.
provider "aws" {
  region = "us-east-1"
}

# Additional provider configuration for west coast region; resources can
# reference this as `aws.west`.
provider "aws" {
  alias  = "west"
  region = "us-west-2"
}
```

</CodeBlockConfig>

<CodeBlockConfig filename="main.tf" hideClipboard>

```hcl
resource "aws_instance" "example" {
  # This "provider" meta-argument selects the AWS provider
  # configuration whose alias is "west", rather than the
  # default configuration.
  provider = aws.west

  # ...
}

module "aws_vpc" {
  source = "./aws_vpc"
  providers = {
    aws = aws.west
  }
}
```

</CodeBlockConfig>

> Any provider block that does not define the `alias` parameter is considered the default provider configuration. We recommend that you always include a default provider configuration and that you define all of your providers in the same file. If you define multiple instances of a provider, define the default instance first. For non-default providers, define the `alias` as the first parameter of the `provider` block.

Reviewer 2: ```suggestion
Any provider block that does not define the `alias` parameter is the default provider configuration. We recommend that you always include a default provider configuration and that you define all of your providers in the same file. If you define multiple instances of a provider, define the default first. For non-default providers, define the `alias` as the first parameter of the `provider` block.
```

Reviewer 2: list?

## Dynamic resource count

> The `for_each` and `count` meta-arguments let you create multiple resources from a single `resource` block depending on run-time conditions. You can use these meta-arguments to make your code flexible and remove the need to duplicate resource blocks. If your instances are almost identical, `count` is appropriate. If some of their arguments need distinct values that cannot be directly derived from an integer, it is safer to use `for_each`.

Reviewer 2: ```suggestion
The `for_each` and `count` meta-arguments let you create multiple resources from a single `resource` block depending on run-time conditions. You can use these meta-arguments to make your code flexible and reduce duplicate resource blocks. If the resources are almost identical, use `count`. If some of arguments need distinct values that you cannot derive from an integer, use `for_each`.
```

> The `for_each` meta-argument accepts a `map` or `set` value, and Terraform will create an instance of that resource for each element in the value you provide. In the following example, Terraform creates an `aws_instance` for each of the strings defined in the `web_instances` variable: "ui", "api", "db" and "metrics". The example uses `each.key` to give each instance a unique name. The `web_private_ips` output uses a [for expression](/terraform/language/expressions/for) to create a map of instance names and their private IP addresses, while the `web_ui_public_ip` output addresses the instance with the key "ui" directly.

Reviewer 2: do we need this context here? is it enough to just generally say that for_each accepts a map input?

<CodeBlockConfig hideClipboard>

> ```hcl

Reviewer 2: now that i see this section in preview, it feels a little long and like this gets in the technical weeds a bit. do you feel we absolutely need these examples here, or are they covered in the docs? anything we can clip?

```

</CodeBlockConfig>

The `count` meta-argument lets you create multiple instances of a resource from a single resource block. In the following example, if the value of `var.web_instance_count` is 2, the code will create two EC2 instances, one named `web_0` and the other named `web_1`.

<CodeBlockConfig hideClipboard>

```hcl
resource "aws_instance" "web" {
  count = var.web_instance_count

  ami           = data.aws_ami.ubuntu.id
  instance_type = "t3.micro"

  tags = {
    Name = "web_${count.index}"
  }
}
```

</CodeBlockConfig>

A common practice to conditionally create resources is to use the `count` meta-argument with a [conditional expression](/terraform/language/expressions/conditionals). In the following example, Terraform will only create the `aws_instance` if `var.enable_metrics` is `true`.

<CodeBlockConfig highlight="8" hideClipboard>

```hcl
variable "enable_metrics" {
  description = "True if the metrics server should be deployed"
  type        = bool
  default     = true
}

resource "aws_instance" "web" {
  count = var.enable_metrics ? 1 : 0

  ami           = data.aws_ami.webapp.id
  instance_type = "t3.micro"
  ##...
}
```

</CodeBlockConfig>

> 
> We recommend that you use meta-arguments in moderation since you sacrifice explicit readability for an implicit understanding of how the configuration works. If the effect of the meta-argument is not immediately obvious, add a comment to make it clear to the reader.

Reviewer 2: ```suggestion

Meta-arguments simplify your code but add complexity, so use them in moderation since. If the effect of the meta-argument is not immediately obvious, use a comment for clarification.
```

Reviewer: @ritsok I'm just a random passerby, but you left the word "since" at the end of the first sentence. 

Also, if you were to ask my opinion, this PR says "use in moderation" in this paragraph and above on line 29 as well, that's a really vague statement, and if you've worked on any moderately scaled Terraform or complex environment you will most likely regularly use for_each. Why should HCP suggest it's users to not to use for_each? Just my opinion though

To learn more about these meta-arguments, refer to the [`for_each`](/terraform/language/meta-arguments/for_each) and [`count`](/terraform/language/meta-arguments/count) documentation.

## .gitignore

Define a `.gitignore` file for your repository to exclude files that you should not publish to version control, such as your state file.

> Always commit:

Reviewer 2: maybe swap the order. feels more logical since the first sentence talks about what to leave out

- All Terraform code files
- Your `.terraform.lock.hcl` [dependency lock file](/terraform/language/files/dependency-lock) 
- A `.gitignore` file that excludes the files listed below
- A `README.md` to describe the code, input variables, and outputs

Exclude:

- Your `terraform.tfstate` state file, including `terraform.tfstate.*` backup state files.
- Your `.terraform.tfstate.lock.info` file. Terraform creates and deletes this file automatically when you run a `terraform apply` command and contains info about your [state lock](/terraform/language/state/locking)
- Your `.terraform` directory, where Terraform downloads providers and child modules. 
[Saved plan files](/terraform/cli/commands/plan#out-filename) that you create when you include the `-out` flat when you run `terraform plan`.
- Any `.tfvars` files that contain sensitive information.

For an example, refer to [GitHub's Terraform .gitignore file](https://github.com/github/gitignore/blob/main/Terraform.gitignore).

> ## Workflow Style

Reviewer 2: sentence case

> Adopting a consistent design style can help with operational stability, security, and make it easier to enforce best practices. The second section of this article covers the design style patterns we recommend that you adopt.

Reviewer 2: ```suggestion
This section reviews standards that enable predictable and secure Terraform workflows, such as:
```

- Pin your Terraform, provider, and module versions.
- Name your module repositories using this three-part name `terraform-<PROVIDER>-<NAME>` when using the Terraform Cloud registry.
- Store local modules at `./modules/<module_name>`.

> - Use the [tfe_outputs](https://registry.terraform.io/providers/hashicorp/tfe/latest/docs/data-sources/outputs) data source or provider-specific data sources to share state between two state files.

Reviewer 2: ```suggestion
- Use the [`tfe_outputs`](https://registry.terraform.io/providers/hashicorp/tfe/latest/docs/data-sources/outputs) data source or provider-specific data sources to share state between two state files.
```

> - Use a secrets manager such as HashiCorp Vault or [dynamic provider credentials](/terraform/tutorials/cloud/dynamic-credentials) when possible so that you do not write secrets to your Terraform code.

Reviewer 2: ```suggestion
- Protect credentials by using [dynamic provider credentials](/terraform/tutorials/cloud/dynamic-credentials) or a secrets manager such as HashiCorp Vault.
```

- Write [tests](/terraform/language/tests) for your modules.

> - Use a linter such as [TFLint](https://github.com/terraform-linters/tflint) to enforce your organization's own coding best practices.

Reviewer 2: this is in the section above

> - Use [policy enforcement](/terraform/cloud-docs/policy-enforcement) on Terraform Cloud to build guardrails for what your developers are and are not allowed to do.

Reviewer 2: ```suggestion
- Use [policy enforcement](/terraform/cloud-docs/policy-enforcement) on Terraform Cloud to set guardrails for infrastructure operations.
```

## Version pinning

To prevent providers and modules upgrades from introducing unintentional changes to your infrastructure, use version pinning. 

> For your providers, define the [required_providers block](/terraform/language/providers/requirements#requiring-providers) inside your `terraform` block to specify which provider version to use. Terraform [version constraints](/terraform/language/providers/requirements#version-constraints) support a range of accepted versions. We recommend that you pin your module to a specific major and minor version as shown in the example below to ensure stability. You can use looser restrictions if you are certain that the module does not introduce breaking changes outside of major version updates.

Reviewer 2: this is mixing both providers and modules

> For your providers, define the [required_providers block](/terraform/language/providers/requirements#requiring-providers) inside your `terraform` block to specify which provider version to use. Terraform [version constraints](/terraform/language/providers/requirements#version-constraints) support a range of accepted versions. We recommend that you pin your module to a specific major and minor version as shown in the example below to ensure stability. You can use looser restrictions if you are certain that the module does not introduce breaking changes outside of major version updates.
> 
> We also recommend that you set a minimum required version of the Terraform binary by setting the `required_version` in your `terraform` block. This ensures that other operators use a version of Terraform that has all of the features your configuration relies on.

Reviewer 2: ```suggestion
For providers, specify provider versions using the [required_providers block](/terraform/language/providers/requirements#requiring-providers). Terraform [version constraints](/terraform/language/providers/requirements#version-constraints) support a range of accepted versions. Pin modules to a specific major and minor version as shown in the example below to ensure stability. You can use looser restrictions if you are certain that the module does not introduce breaking changes outside of major version updates.

We also recommend that you set a minimum required version of the Terraform binary using the `required_version` in your `terraform` block. This requires all operators to use a Terraform version that has all of your configuration's required features.
```

<CodeBlockConfig hideClipboard>

```hcl
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "5.34.0"
    }
  }

>   required_version = ">= 1.7"

Reviewer 2: highlight?

}
```

</CodeBlockConfig>

The above example pins the version of the `hashicorp/aws` provider to version `5.34.0`, and requires that operators use Terraform `1.7` or newer.

For modules sourced from a registry, use the `version` parameter in the `module` block to pin the version. For local modules, Terraform ignores the `version` parameter.

<CodeBlockConfig hideClipboard>

```hcl
module "vault_starter" {
  source  = "hashicorp/vault-starter/aws"
  version = "1.0.0"
  ##...
}
```

</CodeBlockConfig>

## Module repository names

The Terraform registry requires that repositories match a naming convention for all modules that you publish to the registry. Module repositories must use this three-part name `terraform-<PROVIDER>-<NAME>`, where `<NAME>` reflects the type of infrastructure the module manages and `<PROVIDER>` is the main provider the module uses. The `<NAME>` segment can contain additional hyphens, for example, `terraform-google-vault` or `terraform-aws-ec2-instance`.

## Module structure

> Terraform modules allow you to create self-contained, reusable pieces of infrastructure-as-code. You can use modules to abstract away complex code and create multiple resources using a single declaration.
> 
> Use modules to group together logically related resources. For example:

Reviewer 2: ```suggestion
Terraform modules define self-contained, reusable pieces of infrastructure-as-code. 

Use modules to group together logically related resources that you need to provision together. For example:
```

- A networking module that defines a VPC, along with its subnets, gateway, and security groups.
- An application module defining all resources required for each deployment. This stack could include web servers, databases, storage, and supported networking.

Review the [module creation recommended pattern documentation](/terraform/tutorials/modules/pattern-module-creation) and [standard module structure](/terraform/language/modules/develop/structure) for guidance on how to structure your modules.

## Local modules

> Local modules are sourced from local disk rather than a remote module registry. We recommend publishing your modules to a module registry, such as the [Terraform Cloud private module registry](/terraform/cloud-docs/registry). A module registry lets you share modules across your organization for use by other developers. If you cannot use a module registry, organizing part of your code in local modules still reduces the burden of maintaining and updating your code. 

Reviewer 2: ```suggestion
Local modules are sourced from local disk rather than a remote module registry. We recommend publishing your modules to a module registry, such as the [Terraform Cloud private module registry](/terraform/cloud-docs/registry), to easily version, share, and reuse modules across your organization. If you cannot use a module registry, using local modules can simplify maintaining and updating your code. 
```

We recommend that you define child modules in the `./modules/<module_name>` directory. 

## Repository structure

> How you structure your modules and Terraform configuration in version control significantly impacts versioning and operations. In general, we recommend that you store your actual infrastructure configuration separate from your module code.

Reviewer 2: ```suggestion
How you structure your modules and Terraform configuration in version control significantly impacts versioning and operations. We recommend that you store your actual infrastructure configuration separately from your module code.
```

> To structure your modules, we recommend that you store each module in an individual repository. This lets you independently version each module and makes it easier to publish your modules in the private Terraform registry.

Reviewer 2: ```suggestion
Store each module in an individual repository. This lets you independently version each module and makes it easier to publish your modules in the private Terraform registry.
```

## Multiple environments

We recommend that your repository's `main` branch be the source of truth for all environments. For Terraform Cloud and Terraform Enterprise users, we recommend that you use separate workspaces for each environment. For larger codebases, we recommend that you split your resources across multiple workspaces to prevent large state files and limit unintended consequences from changes. For example, you could structure your code as follows:

<CodeBlockConfig hideClipboard>

```
.
├── compute
│   ├── main.tf
│   ├── outputs.tf
│   └── variables.tf
├── database
│   ├── main.tf
│   ├── outputs.tf
│   └── variables.tf
└── networking
    ├── main.tf
    ├── outputs.tf
    └── variables.tf
```

</CodeBlockConfig>

> In this scenario, you would create three workspaces per environment. For example, your production environment would have a "prod-compute", "prod-database", and "prod-networking" workspace. Read more about [Terraform workspace and project best practices](/well-architected-framework/operational-excellence/operational-excellence-workspaces-projects). 

Reviewer 2: ```suggestion
In this scenario, you would create three workspaces per environment. For example, your production environment would have a `prod-compute`, `prod-database`, and `prod-networking` workspace. Read more about [Terraform workspace and project best practices](/well-architected-framework/operational-excellence/operational-excellence-workspaces-projects). 
```

> If you do not use Terraform Cloud or Terraform Enterprise, we recommend that you use modules to encapsulate your configuration, and use a directory for each environment. The configuration in each of these directories would call the local modules, each with parameters specific to their environment. This also lets you maintain separate variable and backend configurations for each environment. 

Reviewer 2: i would draw the connection between directories and state files. also mention that at some point, env-level state files become very risky.

<CodeBlockConfig hideClipboard>

```
├── modules
│   ├── compute
│   │   └── main.tf
│   ├── database
│   │   └── main.tf
│   └── network
│       └── main.tf
├── dev
│   ├── backend.tf
│   ├── main.tf
│   └── variables.tf
├── prod
│   ├── backend.tf
│   ├── main.tf
│   └── variables.tf
└── staging
    ├── backend.tf
    ├── main.tf
    └── variables.tf
```

</CodeBlockConfig>

## State sharing

Since your state contains sensitive information, avoid sharing full state files when possible. 

> If you use Terraform Cloud or Terraform Enterprise and a resource is managed in a separate workspace, use the [`tfe_outputs`](https://registry.terraform.io/providers/hashicorp/tfe/latest/docs/data-sources/outputs) data source to access outputs across workspaces.

Reviewer 2: ```suggestion
If you use Terraform Cloud or Terraform Enterprise and need to reference resources across workspaces, use the [`tfe_outputs`](https://registry.terraform.io/providers/hashicorp/tfe/latest/docs/data-sources/outputs) data source.
```

If you do not use Terraform Cloud or Terraform Enterprise but still need to reference data about other infrastructure resources, use data sources to query the provider. For example, you can use the [`aws_instance` data source](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/instance) to look up an AWS EC2 instance by its ID or tags.

If neither of these options are viable, you can use the [`terraform_remote_state` data source](/terraform/language/state/remote-state-data). If you use this data source with Terraform Cloud or Terraform Enterprise, configure which workspaces have access to your state using the principle of least privilege. Refer to [remote state sharing](/terraform/cloud-docs/workspaces/settings#remote-state-sharing) for more information.

## Secrets management

If you do not configure remote state storage, the Terraform CLI stores the entire state in plaintext on the local disk. State can include sensitive data, such as passwords and private keys. 

> Terraform Cloud and Terraform Enterprise provide state encryption through HashiCorp Vault. 

Reviewer 2: can you make the lead sentence for this list and the CE list next a little more consistent?

> - When using Terraform Enterprise, we recommend that you define and enforce a Sentinel policy to prevent use of the `local_exec` provisioner or external data sources.

Reviewer 2: ```suggestion
- When using Terraform Enterprise, define and enforce a Sentinel policy to prevent use of the `local_exec` provisioner or external data sources.
```
